rewrite-java/src/main/java/org/openrewrite/java/cleanup/CovariantEqualsVisitor.java

```diff
@@ -0,0 +1,132 @@
+
+            /**
+             * Looking for "public boolean equals(EnclosingClassType)" as the method signature match.
+             * We'll replace it with "public boolean equals(Object)"
+             */
+            String ecfqn = enclosingClass.getType().getFullyQualifiedName();
+            if (new MethodMatcher(String.format("%s equals(%s)", ecfqn, ecfqn)).matches(m, enclosingClass) &&
+                    m.hasModifier(J.Modifier.Type.Public) &&
+                    m.getReturnTypeExpr() != null &&
+                    JavaType.Primitive.Boolean.equals(m.getReturnTypeExpr().getType())) {
+
+                if (m.getAnnotations().stream().noneMatch(OVERRIDE_ANNOTATION_SIGNATURE::matches)) {
+                    m = maybeAutoFormat(m,
+                            m.withTemplate(
+                                    template("@Override").build(),
+                                    m.getAnnotations().isEmpty() ? m.getCoordinates().replaceAnnotations() : m.getAnnotations().get(0).getCoordinates().before()
+                            ), p, getCursor().getParentOrThrow());
+                }
+
+                /**
+                 * Change parameter type to Object, and maybe change input parameter name representing the other object.
+                 * This is because we prepend these type-checking replacement statements to the existing "equals(..)" body.
+                 * Therefore we don't want to collide with any existing variable names.
+                 */
+                J.VariableDecls.NamedVar oldParamName = ((J.VariableDecls) m.getParams().iterator().next()).getVars().iterator().next();
+                String paramName = "obj".equals(oldParamName.getSimpleName()) ? "other" : "obj";
+                m = maybeAutoFormat(m,
+                        m.withTemplate(
+                                template("(Object #{})").build(),
+                                m.getCoordinates().replaceParameters(),
+                                paramName
+                        ), p, getCursor().getParentOrThrow());
+
+                /**
+                 * We'll prepend this type-check and type-cast to the beginning of the existing
+                 * equals(..) method body statements, and let the existing equals(..) method definition continue
+                 * with the logic doing what it was doing.
+                 */
+                JavaTemplate equalsBodySnippet = template(EQUALS_BODY_PREFIX_TEMPLATE).build();
+                Object[] params = new Object[]{
+                        paramName,
+                        paramName,
+                        paramName,
+                        enclosingClass.getSimpleName(),
+                        oldParamName.printTrimmed(),
+                        enclosingClass.getSimpleName(),
+                        paramName,
+                        m.getBody().getStatements().stream().map(J::printTrimmed).collect(Collectors.joining(";"))
+                };
```

Comment:
I question whether this templating is the best way to do things. Notice how we're replacing the entire method body declaration, and taking the existing method body statements and appending them to this template. Would appreciate eyes on it to fix this up.

I ran into issues when using the `before` coordinates of the first `Statement` `m.getBody().getStatements().get(0).getCoordinates.before()`

Reply:
Looks reasonable to me. Good work!